### PR TITLE
Fix bugs in L2ZZErrorEstimator

### PR DESCRIPTION
**Pull request by hughcars**
I believe the L2ZZErrorEstimator has two bugs in it:
* It does not make use of the `DofTransformation` pointers to perform `InvTransformPrimal` and `TransformPrimal`, thus is incorrect for ND tets for p>1
* It uses `AddElementVector` to store the flux value into the resulting field. In the case of a zero initialized fully discontinuous field (i.e. L2^d as used for a Poisson flux) this has no effect. But for an ND space this results in double counting. 

The first of these I believe is a plain bug, the second of these I believe to be a bug but could be a misunderstanding on my part. I have also propagated fixing the first issue to what I believe are other instances in `gridfunc.cpp` and `pgridfunc.cpp`. I am not totally sure of these fixes so would appreciate feedback on whether I have misunderstood their application.

This was diagnosed by comparing a flux grid function computed identically to that in the `L2ZZErrorEstimator` function call, to one computed using the CurlInterpolator. The code to construct the interpolated version is
```
using namespace mfem;
mfem::Vector field_vec; // This is populated with the solution vector in ND.

ParGridFunction(&fes); // fes is the FiniteElementSpace associated with field_fes
ParDiscreteLinearOperator curl(&fes, &flux_fes); // ND and RT respectively
curl.AddDomainInterpolator(new CurlInterpolator);
curl.Assemble();
curl.Finalize()
auto *curlop = std::unique_ptr<HypreParMatrix>(curl.ParallelAssemble());

Vector flux(flux_fes.GetTrueVSize());
curlop->Mult(field, flux);

ParGridFunction flux_func(&flux_fes);
flux_func.SetFromTrueDOFs(flux);
```

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3541](https://github.com/mfem/mfem/pull/3541) | @hughcars | @tzanio | @mlstowell + @psocratis + @rw-anderson | 3/25/23 | 4/29/23 | 5/4/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@psocratis and @rw-anderson -- can you please review when you get a chance?

File: fem/pgridfunc.cpp [hunk 1]
```diff
@@ -325,8 +343,8 @@ void ParGridFunction::GetVectorValue(int i, const IntegrationPoint &ip,
    if (nbr_el_no >= 0)
    {
       Array<int> dofs;
-      DofTransformation * doftrans = pfes->GetFaceNbrElementVDofs(nbr_el_no,
-                                                                  dofs);
+      const auto * const doftrans = pfes->GetFaceNbrElementVDofs(nbr_el_no,
+                                                                 dofs);
```
Comment (mlstowell) [inline fem/pgridfunc.cpp:347]: What is the advantage of using `auto` here?

**Comment by mlstowell** [comment]
Hi, @hughcars,

I think your usage of `DofTransformation` is correct here and I thank you for catching these issues. I am curious why you choose to use `const auto * const` rather than `DofTransformation *`. If they are equivalent, I would prefer the second option because it is more specific and more amenable to the use of tools like `grep`.

Regarding the double counting i.e. `AddElementVector` vs. `SetSubVector`, I'm sure you are correct that this double counting is a problem. However, I'm not sure of the appropriate fix. Switching to `SetSubVector` will remove the double counting but it introduces a dependence on the order of element traversal. I think what we want is to build a right-hand-side vector `b` by summing contributions from neighboring elements. I'm not sure why the code was not written this way initially (@rw-anderson or @psocratis may have keener insight on this matter.) 

My inclination is to rewrite this portion of `L2ZZErrorEstimator` but your fix is an improvement over the double counting error that currently exists.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Regarding the use of `auto`, I don't think the quoted C++ guidelines apply in this case. There is no repetition of the typename and the function call doesn't give any clues to the reader indicating what type should be expected. This would require the reader to lookup the function in the header files to figure out what object is being used. Unless the other reviewers disagree I would prefer using `DofTransformation *` (adding the `const` keywords is fine).

I see your point about the flux computation. Without coefficients and assuming the user chooses the correct space for the flux everything should be the same regardless of traversal order.

**Comment by hughcars** [comment]
> Regarding the use of `auto`, I don't think the quoted C++ guidelines apply in this case. There is no repetition of the typename and the function call doesn't give any clues to the reader indicating what type should be expected. This would require the reader to lookup the function in the header files to figure out what object is being used. Unless the other reviewers disagree I would prefer using `DofTransformation *` (adding the `const` keywords is fine).
> 
> I see your point about the flux computation. Without coefficients and assuming the user chooses the correct space for the flux everything should be the same regardless of traversal order.

Fixed in https://github.com/mfem/mfem/pull/3541/commits/1c25aeb978448c354da5140aad68c52c4ed4c364

**Review by mlstowell — APPROVED** [review]

**Review by psocratis — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by hughcars** [comment]
> Can we mention in `CHANGELOG` so users are not surprised to see changes?

Yes, I can write an entry, does this work?
```
- Bugfix for ZZ error estimation with Nedelec spaces: interface unknowns are no longer double counted, and element transformations for p >= 2 Nedelec elements are now accounted for correctly.
```

**Pull request merged**